src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs

```diff
@@ -98,75 +104,114 @@ public async Task<DownloadResult> DownloadPackageAsync(InstallRequest installReq
-            IEnumerable<string> attemptedSources = sources != null ? sources : new[] { PublicNuGetFeed };
-            foreach (string source in attemptedSources)
+            SourceCacheContext cache = new SourceCacheContext();
+            Dictionary<PackageSource, PackageMetadataResource> resources = new Dictionary<PackageSource, PackageMetadataResource>();
+            foreach (PackageSource source in packageSources)
             {
                 SourceRepository repository = Repository.Factory.GetCoreV3(source);
                 try
                 {
-                    PackageMetadataResource resource = await repository.GetResourceAsync<PackageMetadataResource>().ConfigureAwait(false);
+                    PackageMetadataResource resource = await repository.GetResourceAsync<PackageMetadataResource>(cancellationToken).ConfigureAwait(false);
                     resources[source] = resource;
                 }
                 catch (Exception e)
                 {
-                    _nugetLogger.LogError($"Failed to load the NuGet source {source}");
-                    _nugetLogger.LogDebug($"Reason: {e.Message}");
+                    _nugetLogger.LogError($"Failed to load the NuGet source {source}.");
+                    _nugetLogger.LogDebug($"Reason: {e.Message}.");
                 }
             }
 
             if (!resources.Any())
             {
-                throw new InvalidNuGetSourceException("Failed to load NuGet sources", attemptedSources);
+                throw new InvalidNuGetSourceException("Failed to load NuGet sources", packageSources.Select(source => source.Source));
             }
 
-            ConcurrentDictionary<string, IEnumerable<IPackageSearchMetadata>> searchResults = new ConcurrentDictionary<string, IEnumerable<IPackageSearchMetadata>>();
-            foreach (KeyValuePair<string, PackageMetadataResource> resource in resources)
+            Dictionary<PackageSource, IEnumerable<IPackageSearchMetadata>> searchResults = new Dictionary<PackageSource, IEnumerable<IPackageSearchMetadata>>();
+            foreach (KeyValuePair<PackageSource, PackageMetadataResource> resource in resources)
             {
-                IEnumerable<IPackageSearchMetadata> foundPackages = await resource.Value.GetMetadataAsync(
-                    packageIdentifier,
-                    includePrerelease: false,
-                    includeUnlisted: false,
-                    cache,
-                    _nugetLogger,
-                    cancellationToken).ConfigureAwait(false);
-                if (foundPackages.Any())
+                try
                 {
-                    _nugetLogger.LogVerbose($"Found {foundPackages.Count()} {packageIdentifier} packages in NuGet feed {resource.Key}");
-                    searchResults[resource.Key] = foundPackages;
+                    IEnumerable<IPackageSearchMetadata> foundPackages = await resource.Value.GetMetadataAsync(
```

Comment:
It would be nice to do these in parallel. Checking 3 different sources will take roughly 3 times longer compared to checking a single source. Same goes for  `GetResourceAsync` in the loop above.